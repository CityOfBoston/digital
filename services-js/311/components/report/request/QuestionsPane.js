// @flow

import React from 'react';
import { css } from 'glamor';

import AttributeField from './AttributeField';

import type { Service } from '../../../data/types';
import type { State as Request } from '../../../data/store/request';

export type ExternalProps = {
  service: Service,
  nextFunc: () => void,
}

export type ValueProps = {
  request: Request,
}

export type ActionProps = {
  setAttribute: (string, string | string[]) => void,
}

export type Props = ExternalProps & ValueProps & ActionProps;

const DESCRIPTION_STYLE = css({
  fontFamily: '"Lora", Georgia, serif',
  fontSize: 28,
  fontStyle: 'italic',
  lineHeight: 1.7,
});

const ADDRESS_STYLE = css({
  textTransform: 'uppercase',
  whiteSpace: 'pre-line',
  margin: '30px 0 0',
  fontSize: 16,
  fontWeight: 'bold',
});

function selectAttributes(service) {
  if (service.metadata && service.metadata.attributes) {
    return service.metadata.attributes;
  } else {
    return [];
  }
}

function renderDescription({ description }: Request) {
  if (description.length) {
    return <div className={DESCRIPTION_STYLE}>“{description}”</div>;
  } else {
    return null;
  }
}

function renderLocation({ address }: Request) {
  return <div className={ADDRESS_STYLE}>{address}</div>;
}

export default function QuestionsPane({ service, request, setAttribute, nextFunc }: Props) {
  // TODO(finneganh): required fields
  const allowSubmit = true;

  return (
    <div>
      { renderDescription(request) }
      { renderLocation(request) }

      {
        selectAttributes(service).map((a) => (
          <AttributeField attribute={a} key={a.code} attributeChanged={setAttribute} currentValue={request.attributes[a.code]} />
        ))
      }

      <button onClick={nextFunc} disabled={!allowSubmit}>Next</button>
    </div>
  );
}
