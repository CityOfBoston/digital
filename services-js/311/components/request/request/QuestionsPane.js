// @flow

import React from 'react';
import { css } from 'glamor';
import { action, extras } from 'mobx';
import { observer } from 'mobx-react';
import Dropzone from 'react-dropzone';

import SectionHeader from '../../common/SectionHeader';
import { MEDIA_LARGE } from '../../style-constants';
import AttributeField from './AttributeField';

import CloudinaryImageUpload from '../../../data/external/CloudinaryImageUpload';

import type { AppStore } from '../../../data/store';
import type RequestForm from '../../../data/store/RequestForm';

export type Props = {|
  store: AppStore,
  requestForm: RequestForm,
  serviceName: string,
  serviceDescription: ?string,
  nextFunc: () => mixed,
  nextIsSubmit: boolean,
|};

const DROPZONE_STYLE = css({
  width: '100%',
  border: 'none',
  position: 'relative',
});

const DRAG_RING_STYLE = css({
  borderStyle: 'dashed',
  borderColor: 'white',
  position: 'absolute',
  top: '1rem',
  left: '1rem',
  bottom: '1rem',
  right: '1rem',
});

const BOTTOM_ROW_STYLE = css({
  alignItems: 'center',
});

const RIGHT_ON_LARGE_STYLE = css({
  [MEDIA_LARGE]: {
    textAlign: 'right',
  },
});

const PROGRESS_STYLE = css({
  position: 'absolute',
  left: 0,
  bottom: 0,
  height: 6,
  backgroundColor: '#288be4',
});

@observer
export default class QuestionsPane extends React.Component<Props> {
  imageUploader: CloudinaryImageUpload = new CloudinaryImageUpload();
  dropEl: ?Dropzone;

  @action
  componentWillMount() {
    const { store, requestForm } = this.props;
    this.imageUploader.config = store.apiKeys.cloudinary;
    this.imageUploader.adoptedUrlObservable = extras.getAtom(
      requestForm,
      'mediaUrl'
    );
  }

  @action
  componentWillReceiveProps(newProps: Props) {
    const { store, requestForm } = newProps;
    this.imageUploader.config = store.apiKeys.cloudinary;
    this.imageUploader.adoptedUrlObservable = extras.getAtom(
      requestForm,
      'mediaUrl'
    );
  }

  @action
  componentWillUnmount() {
    this.imageUploader.adoptedUrlObservable = null;
  }

  @action.bound
  handleDrop(acceptedFiles: File[]) {
    if (acceptedFiles[0]) {
      this.imageUploader.upload(acceptedFiles[0]);
    }
  }

  @action.bound
  handleUploadPhoto() {
    if (this.dropEl) {
      this.dropEl.open();
    }
  }

  @action.bound
  handleRemoveImage() {
    this.imageUploader.remove();
  }

  @action.bound
  handleUpdateDescription(ev: SyntheticInputEvent<>) {
    const { requestForm } = this.props;
    requestForm.description = ev.target.value;
  }

  @action.bound
  handleSubmit(ev: SyntheticInputEvent<>) {
    const { nextFunc } = this.props;

    ev.preventDefault();

    nextFunc();
  }

  setDropEl = (dropEl: ?Dropzone) => {
    this.dropEl = dropEl;
  };

  render() {
    const {
      requestForm,
      serviceName,
      serviceDescription,
      nextIsSubmit,
    } = this.props;
    const { description, questions, questionRequirementsMet } = requestForm;

    const questionsEls = [];
    questions.forEach((q, i) => {
      if (!q.visible) {
        return;
      }

      questionsEls.push(
        <div key={q.code}>
          <AttributeField question={q} />
        </div>
      );

      if (i < questions.length - 1) {
        questionsEls.push(
          <hr className="hr hr--dash m-v500" aria-hidden key={`${q.code}-HR`} />
        );
      }
    });

    return (
      <form onSubmit={this.handleSubmit}>
        <SectionHeader>
          {serviceName}
        </SectionHeader>

        <div className="t--intro m-v300">
          {serviceDescription}
        </div>

        <div className="g g--top">
          <div className="g--7 m-v500">
            <div className="txt">
              <label
                className="txt-l"
                htmlFor="QuestionsPane-description"
                style={{ marginTop: 0 }}
              >
                Description
              </label>
              <textarea
                name="description"
                id="QuestionsPane-description"
                className="txt-f"
                value={description}
                placeholder="Please describe your request"
                onChange={this.handleUpdateDescription}
                rows="5"
              />
            </div>

            {questionsEls}
          </div>

          {this.renderImageUpload()}
        </div>

        <div className={`g m-v500 ${BOTTOM_ROW_STYLE.toString()}`}>
          <div
            className={`g--9 t--info m-v200 ${RIGHT_ON_LARGE_STYLE.toString()}`}
          >
            {!questionRequirementsMet &&
              <span>
                Please fill out <span className="t--req">required</span> fields
                to continue
              </span>}
          </div>
          <button
            className="btn g--3"
            type="submit"
            disabled={!questionRequirementsMet}
          >
            {nextIsSubmit ? 'Submit Request' : 'Next'}
          </button>
        </div>
      </form>
    );
  }

  renderImageUpload() {
    const {
      errorMessage,
      displayUrl,
      uploading,
      uploadingProgress,
      canRemove,
    } = this.imageUploader;

    let buttonAction = null;
    let buttonTitle = null;
    let buttonClasses = 'btn btn--w btn--b';

    if (uploading) {
      buttonAction = this.handleRemoveImage;
      buttonTitle = 'Cancel Upload';
      buttonClasses = `${buttonClasses} btn--r-hov`;
    } else if (canRemove) {
      buttonAction = this.handleRemoveImage;
      buttonTitle = 'Remove Photo';
    } else {
      buttonAction = this.handleUploadPhoto;
      buttonTitle = 'Upload Photo';
    }

    return (
      <div className="g--5 m-v500">
        <div className="br br-a200">
          <Dropzone
            className={DROPZONE_STYLE.toString()}
            ref={this.setDropEl}
            onDrop={this.handleDrop}
            multiple={false}
            accept="image/*"
          >
            {({ isDragActive }) => {
              const out = [
                displayUrl
                  ? <img
                      key="img"
                      style={{ display: 'block', width: '100%' }}
                      alt=""
                      aria-hidden
                      src={displayUrl}
                    />
                  : <img
                      key="img"
                      style={{
                        display: 'block',
                        width: '100%',
                        height: 'auto',
                      }}
                      aria-hidden
                      alt=""
                      width="479"
                      height="324"
                      src="/assets/img/image-upload.png"
                    />,
              ];

              if (isDragActive) {
                out.push(
                  <div
                    key="ring"
                    className={`br-a300 ${DRAG_RING_STYLE.toString()}`}
                  />
                );
              }

              return out;
            }}
          </Dropzone>

          {errorMessage &&
            <div className="t--info br br-t200 p-a300">
              {errorMessage}
            </div>}

          <div className="br br-t200" style={{ position: 'relative' }}>
            <button
              className={buttonClasses}
              style={{
                visibility: buttonAction ? 'visible' : 'hidden',
                width: '100%',
              }}
              type="button"
              onClick={buttonAction}
            >
              {buttonTitle}
            </button>
            {uploading &&
              <div
                className={PROGRESS_STYLE}
                style={{ width: `${100 * uploadingProgress}%` }}
              />}
          </div>
        </div>
      </div>
    );
  }
}
