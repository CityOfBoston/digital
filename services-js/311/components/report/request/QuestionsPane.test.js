// @flow

import React from 'react';
import renderer from 'react-test-renderer';

import type { Service, Request } from '../../../data/types';

import QuestionsPane from './QuestionsPane';

export const DEFAULT_SERVICE: Service = {
  name: 'Cosmic Incursion',
  code: 'CSMCINC',
  hasMetadata: false,
  metadata: null,
};

export const DEFAULT_REQUEST: Request = {
  code: null,
  description: '',
  firstName: '',
  lastName: '',
  email: '',
  phone: '',
  location: null,
  address: '',
  attributes: {},
};

export const SERVICE_WITH_METADATA: Service = {
  name: 'Cosmic Incursion',
  code: 'CSMCINC',
  hasMetadata: true,
  metadata: {
    attributes: [{
      required: false,
      type: 'TEXT',
      code: 'ST-CMTS',
      description: 'Please provide any other relevant information:',
      values: null,
    }],
  },
};

export const REQUEST_WITH_METADATA: Request = {
  code: null,
  description: '',
  firstName: '',
  lastName: '',
  email: '',
  phone: '',
  location: null,
  address: '',
  attributes: {
    ST_CMTS: 'A portal to the Negative Zone has opened',
  },
};

const ACTIONS = {
  nextFunc: jest.fn(),
  setAttribute: jest.fn(),
};

test('blank request', () => {
  const component = renderer.create(
    <QuestionsPane service={DEFAULT_SERVICE} request={DEFAULT_REQUEST} {...ACTIONS} loopbackGraphql={jest.fn()} />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('existing description', () => {
  const component = renderer.create(
    <QuestionsPane
      service={DEFAULT_SERVICE}
      request={{ ...DEFAULT_REQUEST, description: 'Please pick up my bulk items. ' }}
      {...ACTIONS}
      loopbackGraphql={jest.fn()}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('service with metadata', () => {
  const component = renderer.create(
    <QuestionsPane
      service={SERVICE_WITH_METADATA}
      request={REQUEST_WITH_METADATA}
      {...ACTIONS}
      loopbackGraphql={jest.fn()}
    />,
  );

  expect(component.toJSON()).toMatchSnapshot();
});
