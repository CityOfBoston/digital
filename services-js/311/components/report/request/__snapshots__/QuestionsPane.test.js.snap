exports[`test blank request 1`] = `
<div>
  <div
    className={
      Object {
        "data-css-dgukzt": "",
      }
    }>
    
  </div>
  <button
    disabled={false}
    onClick={[Function]}>
    Next
  </button>
</div>
`;

exports[`test existing description 1`] = `
<div>
  <div
    className={
      Object {
        "data-css-wb46si": "",
      }
    }>
    “
    Please pick up my bulk items. 
    ”
  </div>
  <div
    className={
      Object {
        "data-css-dgukzt": "",
      }
    }>
    
  </div>
  <button
    disabled={false}
    onClick={[Function]}>
    Next
  </button>
</div>
`;

exports[`test service with metadata 1`] = `
<div>
  <div
    className={
      Object {
        "data-css-dgukzt": "",
      }
    }>
    
  </div>
  <label>
    <p>
      Please provide any other relevant information:
       
    </p>
    <textarea
      className={
        Object {
          "data-css-1bm19ha": "",
        }
      }
      name="ST-CMTS"
      onChange={[Function]}
      value={undefined} />
  </label>
  <button
    disabled={false}
    onClick={[Function]}>
    Next
  </button>
</div>
`;
